ingest-app/src/cmr/ingest/api/subscriptions.clj

```diff
@@ -124,20 +124,73 @@
       csk/->snake_case
       (str "_" (UUID/randomUUID))))
 
-(defn ingest-subscription
-  "Processes a request to create or update a subscription. Note, this will allow
-  unlimited subscriptions to be created by users for themselves. Revisit if this
-  Becomes a problem"
-  [provider-id opt-native-id request]
+(defn native-id-collision?
+  "Queries metadata db for a matching provider-id and native-id pair."
+  [context provider-id native-id]
+  (let [query {:provider-id provider-id
+               :native-id native-id
+               :exclude-metadata true
+               :latest true}]
+    (-> context
+        (mdb/find-concepts query :subscription)
+        seq)))
+
+(defn get-unique-native-id
+  "Get a native-id that is unique by testing against the database."
+  [context subscription]
+  (let [native-id (generate-native-id subscription)
+        provider-id (:provider-id subscription)]
+    (if (native-id-collision? context provider-id native-id)
+      (do
+        (warn (format "Collision detected while generating native-id [%s] for provider [%s], retrying."
+                      native-id provider-id))
+        (get-unique-native-id context subscription))
+      native-id)))
+
+(defn create-subscription
+  "Processes a request to create a subscription. A native id will be generated."
+  [provider-id request]
   (let [{:keys [body content-type headers request-context]} request]
     (common-ingest-checks request-context provider-id)
     (let [tmp-concept (api-core/body->concept!
                        :subscription provider-id (str (UUID/randomUUID)) body content-type headers)
-          native-id (or opt-native-id (generate-native-id tmp-concept))
+          native-id (get-unique-native-id request-context tmp-concept)
           concept (assoc tmp-concept :native-id native-id)]
       (check-subscription-ingest-permission request-context concept provider-id)
```

Comment:
Since we know there is no existing concept when creating a subscription, there is no need to execute the find-concept call inside `check-subscription-ingest-permission` function. Refactor `check-subscription-ingest-permission` function so that we can bypass the unnecessary find-concept call.